lib/modules/manager/poetry/schema.ts

```diff
@@ -1,17 +1,152 @@
+import * as poetryVersioning from '../../versioning/poetry';
+import type { PackageDependency } from '../types';
 
-const PoetryDependencySchema = z.object({
-  path: z.string().optional(),
-  git: z.string().optional(),
-  tag: z.string().optional(),
-  version: z.string().optional(),
-});
+const PoetryPathDependency = z
+  .object({
+    path: z.string(),
+    version: z.string().optional().catch(undefined),
+  })
+  .transform(({ version }): PackageDependency => {
+    const dep: PackageDependency = {
+      datasource: PypiDatasource.id,
+      skipReason: 'path-dependency',
+    };
+
+    if (version) {
+      dep.currentValue = version;
+    }
+
+    return dep;
+  });
+
+const PoetryGitDependency = z
+  .object({
+    git: z.string(),
+    tag: z.string().optional().catch(undefined),
+    version: z.string().optional().catch(undefined),
+  })
+  .transform(({ git, tag, version }): PackageDependency => {
+    if (!tag) {
+      const res: PackageDependency = {
+        datasource: GitRefsDatasource.id,
+        packageName: git,
+        skipReason: 'git-dependency',
+      };
+
+      if (version) {
+        res.currentValue = version;
+      }
+
+      return res;
+    }
+
+    const parsedUrl = parseGitUrl(git);
+    if (parsedUrl.source !== 'github.com') {
+      return {
+        datasource: GitRefsDatasource.id,
+        currentValue: tag,
+        packageName: git,
+        skipReason: 'git-dependency',
+      };
+    }
+
+    const { owner, name } = parsedUrl;
+    const repo = `${owner}/${name}`;
+    return {
+      datasource: GithubTagsDatasource.id,
+      currentValue: tag,
+      packageName: repo,
+    };
+  });
+
+const PoetryPypiDependency = z.union([
+  z
+    .object({ version: z.string().optional() })
+    .transform(({ version: currentValue }): PackageDependency => {
+      if (!currentValue) {
+        return { datasource: PypiDatasource.id };
+      }
+
+      return {
+        datasource: PypiDatasource.id,
+        managerData: { nestedVersion: true },
+        currentValue,
+      };
+    }),
+  z.string().transform(
+    (version): PackageDependency => ({
+      datasource: PypiDatasource.id,
+      currentValue: version,
+      managerData: { nestedVersion: false },
+    })
+  ),
+]);
+
+const PoetryDependencySchema = z.union([
+  PoetryPathDependency,
+  PoetryGitDependency,
+  PoetryPypiDependency,
+]);
+
+const PoetryArraySchema = z.array(z.unknown()).transform(
+  (): PackageDependency => ({
+    datasource: PypiDatasource.id,
+    skipReason: 'multiple-constraint-dep',
+  })
+);
+
+const PoetryValue = z.union([PoetryDependencySchema, PoetryArraySchema]);
+type PoetryValue = z.infer<typeof PoetryValue>;
 
 export const PoetryDependencyRecord = LooseRecord(
   z.string(),
-  z.union([PoetryDependencySchema, z.array(PoetryDependencySchema), z.string()])
-);
+  PoetryValue.transform((dep) => {
+    if (dep.skipReason) {
+      return dep;
+    }
+
+    // istanbul ignore if: normaly should not happen
```

Comment:
why? is the skip reason normally already be set before?

Reply:
Exactly